### PR TITLE
Allow index pattern change in visualizations

### LINKED ISSUE #707: Change the index pattern on a visualization
One of the most requested features in Kibana was the ability to [change the index pattern on a visualisation](https://github.com/elastic/kibana/issues/17542).  This issue was eventually closed as it was implemented in Kibana's Lens feature.

Currently, once a visualisation has been made then it is impossible to change the index pattern unless you go into "Saved Objects" and manually alter the visualisation JSON, which is cumbersome, open to all sorts of issues and definitely not for casual users.

It would be useful to provide some simple mechanism to allow users to alter the index pattern once a chart has been created - maybe similar to how index patterns can be changed in the Discover tab.

(Also see https://github.com/elastic/kibana/issues/3668 for additional discussion)

### PR DESCRIPTION
**Pull request by AMoo-Miki**
Signed-off-by: Miki <miki@amazon.com>

### Description
* Introduced VisIndexPatternSelector: A plugin that offers the UX for modifying index patterns to other plugins
* Modified the IndexPatternManagement plugin to export `IndexPatternManagmentContext` type
* Modified `DefaultEditorSideBar` to use the new plugin to allow changing the index pattern of visualizations that use it
* Modified `SerializedVisData` to include information about the index pattern (required for `DefaultEditorSideBar` integration)

https://user-images.githubusercontent.com/3527403/148879708-d867df73-4c3d-48c0-bae1-f1d1fb546a2e.mp4

 
### Issues Resolved
#707 
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
    - [-] `yarn test:jest`
    - [-] `yarn test:jest_integration`
    - [ ] `yarn test:ftr`
- [X] New functionality has been documented.
- [X] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/vis_index_pattern_selector/public/components/change_indexpattern.test.tsx [hunk 1]
```diff
@@ -0,0 +1,79 @@
+/*
```
Comment (tmarkley) [inline src/plugins/vis_index_pattern_selector/public/components/change_indexpattern.test.tsx:1]: We can use the simplified header for new files:

https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/.eslintrc.js#L36-L41

Comment (AMoo-Miki) [reply src/plugins/vis_index_pattern_selector/public/components/change_indexpattern.test.tsx:1]: Good catch.

**Comment by tmarkley** [comment]
@AMoo-Miki what's left for this PR? Is it ready for review?

**Comment by seanneumann** [comment]
Talked to @ashwin-pc. We believe this is working, but need to follow-up with UX and PM to see if this is a feature we need.

**Comment by kgcreative** [comment]
Yes -- being able to change your data source will be important. Today, the only way to do that is to manually edit the saved object. I think the main challenge would be matching settings after you switch data source. (What happens when the schema is different between index patterns)

**Comment by AMoo-Miki** [comment]
A ton of changes have been made since this PR was raised. We will need to re-test this before merging it.

**Comment by ashwin-pc** [comment]
@AMoo-Miki Can we close this then and reopen it when we have time to prioritize this?

**Comment by ashwin-pc** [comment]
Closing it since its been in the backlog for a year and product roadmap has changed since.

**Pull request closed without merging**